### PR TITLE
examples: Add notebooks to CI, add python3.11 in testing

### PR DESCRIPTION
**Pull request by georgebisbas**

### CODE REVIEW

**Review by webmiche — COMMENTED** [review]
This looks fine, IMO. I think we should have a broader discussion whether we actually want this to be a CI job that needs to be fulfilled, as I fear that it might have the effect that refactoring PRs become bigger as this will need to be updated as well.

I am not familiar enough with github, but a "good to have" class of CI jobs might make sense here. Then we can merge refactorings without having to update the notebooks while still being reminded that we need to update the notebooks before the next release.

File: xdsl/dialects/cmath.py [hunk 1]
```diff
@@ -2,11 +2,10 @@
 from dataclasses import dataclass
 
 from xdsl.dialects.builtin import Float32Type, Float64Type
-from xdsl.ir import MLContext, MLIRType
+from xdsl.ir import MLContext, MLIRType, ParametrizedAttribute
```
[the remark this answers is not in the document]
Comment (georgebisbas) [reply xdsl/dialects/cmath.py:5]: Yes, while doing this PR I found out a few wrong imports and fixed them

**Comment by georgebisbas** [comment]
> This looks fine, IMO. I think we should have a broader discussion whether we actually want this to be a CI job that needs to be fulfilled, as I fear that it might have the effect that refactoring PRs become bigger as this will need to be updated as well.
> 
> I am not familiar enough with github, but a "good to have" class of CI jobs might make sense here. Then we can merge refactorings without having to update the notebooks while still being reminded that we need to update the notebooks before the next release.

Do you mean that I should add the notebooks as a different github workflow?
I guess we totally agree that notebooks should be under CI and be edited accordingly if their output is changed

**Comment by webmiche** [comment]
I think they should be a different workflow, yes.

I am arguing about whether it makes sense to make our PRs more cluttered with notebook changes. But I agree, if we want to have consistent documentation, there is probably no way around that.

**Comment by georgebisbas** [comment]
> I think they should be a different workflow, yes.

yes

**Comment by superlopuh** [comment]
Very cool! If we're committed to having a low barrier of entry to new users then not breaking the notebooks that currently function as the best documentation we have is a good commitment, even if they add complexity to PRs, in my opinion

**Review by math-fehr — APPROVED** [review]
Perfect!
I added two nits comments on the CI file, but otherwise that looks good to me.
I'm more in the same side as George and Sasha, I prefer having the notebooks always updated, since this will probably be the preferred way of learning xDSL.

**Comment by webmiche** [comment]
Maybe to clarify a bit: I am also in favor of adding the notebooks to the CI. The thing I was thinking about was having two classes of CI jobs: 1. necessary ones, that have to pass on every commit, 2. optional ones, that have to pass before we release a new version.

Then, I would say that our testing needs to be in 1., whereas documentation updates should bin in category 2, IMO.

If git does not support such functionality, I am also in favor of just adding the notebooks as a normal CI job.

**Comment by math-fehr** [comment]
I see, so my question would be: How do we expect users to use the notebooks.
If we expect them to `pip install xdsl`, then I agree with you, we do not have to update it every single time (though I would personally prefer doing so).
However, if we expect them to `pip install -e .`, then we need the notebooks to always be updated, since they will be using the current `xdsl` commit.

**Comment by webmiche** [comment]
Okay, yes I agree. I expected users to only use releases and not each commit. If we want them to be able to use each commit, we have to update the notebooks, yes.

File: .github/workflows/ci-notebooks.yml [hunk 2]
```diff
@@ -0,0 +1,39 @@
+# This workflow will install Python dependencies, run tests and lint with a single version of Python
+# For more information see: https://help.github.com/actions/language-and-framework-guides/using-python-with-github-actions
+
+name: CI notebooks
```
[the remark this answers is not in the document]
Comment (georgebisbas) [reply .github/workflows/ci-notebooks.yml:4]: yes

File: .github/workflows/ci-notebooks.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      with:
+        python-version: ${{ matrix.python-version }}
+    - name: Upgrade pip
+      run: |
+        pip install --upgrade pip
+    - name: Install the package locally
+      run: pip install -e .
+    - name: Run examples/tutorials
+      run: |
+        pip install nbval
```
[the remark this answers is not in the document]
Comment (georgebisbas) [reply .github/workflows/ci-notebooks.yml:37]: yes

**Comment by georgebisbas** [comment]
> Maybe to clarify a bit: I am also in favor of adding the notebooks to the CI. The thing I was thinking about was having two classes of CI jobs: 1. necessary ones, that have to pass on every commit, 2. optional ones, that have to pass before we release a new version.
> 
> Then, I would say that our testing needs to be in 1., whereas documentation updates should bin in category 2, IMO.
> 
> If git does not support such functionality, I am also in favor of just adding the notebooks as a normal CI job.

Everything so far in notebooks is related to "printing stuff" which should be kept consistent.
YOu would like to avoid, having users complaining that cell-5 gives me something new.
So far it is only two notebooks, I do not see the reason to split in optional or not.

**Comment by math-fehr** [comment]
That looks good to me! Can we merge it?

**Review by webmiche — APPROVED** [review]

**Comment by georgebisbas** [comment]
Sure!

**Pull request merged**